java/src/test/java/ai/onnxruntime/OnnxTensorTest.java

```diff
@@ -88,6 +88,91 @@ public void testScalarCreation() throws OrtException {
     }
   }
 
+  @Test
+  public void testBufferCreation() throws OrtException {
+    OrtEnvironment env = OrtEnvironment.getEnvironment();
+
+    // Test creating a value from an array
+    // Arrays result in tensors allocated by ORT, so they do not have a backing java.nio.Buffer
+    float[] arrValues = new float[] {0, 1, 2, 3, 4};
+    try (OnnxTensor t = OnnxTensor.createTensor(env, arrValues)) {
+      // array creation isn't backed by buffers
+      Assertions.assertFalse(t.ownsBuffer());
+      Assertions.assertFalse(t.getBufferRef().isPresent());
+      FloatBuffer buf = t.getFloatBuffer();
+      float[] output = new float[arrValues.length];
+      buf.get(output);
+      Assertions.assertArrayEquals(arrValues, output);
+
+      // Can't modify the tensor through this buffer.
+      buf.put(0, 25);
+      Assertions.assertArrayEquals(arrValues, output);
+    }
+
+    // Test creating a value from a non-direct byte buffer
+    // Non-direct byte buffers are allocated on the Java heap and must be copied into off-heap
+    // direct byte buffers
+    // which can be directly passed to ORT
+    FloatBuffer nonDirectBuffer = FloatBuffer.allocate(5);
+    nonDirectBuffer.put(arrValues);
+    nonDirectBuffer.rewind();
+    try (OnnxTensor t = OnnxTensor.createTensor(env, nonDirectBuffer, new long[] {1, 5})) {
+      // non-direct buffers trigger a copy
+      Assertions.assertTrue(t.ownsBuffer());
+      // tensors backed by buffers can get the buffer ref back out
+      Assertions.assertTrue(t.getBufferRef().isPresent());
+      FloatBuffer buf = t.getFloatBuffer();
+      float[] output = new float[arrValues.length];
+      buf.get(output);
+      Assertions.assertArrayEquals(arrValues, output);
+
+      // Can't modify the tensor through getFloatBuffer.
+      buf.put(0, 25);
+      Assertions.assertArrayEquals(arrValues, output);
+
+      // Can modify the tensor through getBufferRef.
+      FloatBuffer ref = (FloatBuffer) t.getBufferRef().get();
+      ref.put(0, 25);
+      buf = t.getFloatBuffer();
+      buf.get(output);
+      Assertions.assertEquals(25, output[0]);
+    }
+
+    // Test creating a value from a direct byte buffer
+    // Direct byte buffers can be passed into ORT without additional copies or processing
+    FloatBuffer directBuffer =
+        ByteBuffer.allocateDirect(5 * 4).order(ByteOrder.nativeOrder()).asFloatBuffer();
+    directBuffer.put(arrValues);
```

Comment:
Does this `put()` copies data from managed buffer?

Reply:
That `put()` call copies the data from the supplied `arrValues` array into `directBuffer`.

Reply:
My understanding managed code typically would manipulate data using managed arrays. Then, ideally, we want to avoid the copy and let native code use it as input.

Alternatively, the user should be able to allocate a direct buffer and then manipulate the data there, but that's not usually the case. People typically use arrays to do so.

So I am not seeing that the intended behavior is implemented.

In C# we either pass managed buffer directly to the native code as input or we read the native buffer as an output. All of that is accomplished using the same interfaces.

Same in C++.

Here, the user should copy data in either case, direct or non-direct buffer. IMHO, the interface that takes non-direct buffer should not exist, as I pointed out in my other comment. Its purpose is even less clear.

Reply:
I think this is due to the difference in managed behaviour between C# and Java. This is idiomatic for Java. A direct byte buffer is both managed (as Java will deallocate it when it goes out of scope) and accessible from native code. It's not possible to prevent it from taking a non-direct buffer using the type system if it accepts a direct buffer, they are the same type. It's also not possible to expose a Java array directly to native code without pinning it causing GC issues and we can't do multidimensional arrays like that as they aren't densely stored in Java.

The test copies data into the buffer as I'd like the test to use the same array for all the tests. Users can create memory mapped direct byte buffers (https://docs.oracle.com/javase/8/docs/api/java/nio/MappedByteBuffer.html), but I didn't do that for tests as this is equivalent in terms of the tested functionality.

Reply:
The differences are noted, but we care about reducing the number of copies and the clarity. I am certainly not advocating multi-dim array. ORT never had such a notion, and our interfaces do not treat it as such. (DenseTensor had multi-dim accessors, but that's different).

I am worried that:

1. There not a true direct buffer access.
2. API confusion.

So I am questioning if we need this at all, since the user end's up copying the data.

Reply:
There is true direct buffer access. Tensors created backed by a direct byte buffer have their elements individually addressable on the Java side and the pointer is directly accessible for ORT on the native side. Users have to write values into the buffer somehow, in this test case we copy them from the array of values used in all the other test cases. In other cases they'd write in the token ids or pixel values from wherever they arrive. In the stable diffusion example the direct buffer is operated on to apply the scheduler's update and then passed right back into a `run` call.